### PR TITLE
TWCC based congestion control - v0

### PR DESCRIPTION
**Pull request by boks1971**
Making a PR to kick off this development. It is missing some bits (see notes below), but wanted to get this going so that we can make incremental/iterative commits. This is already quite large a change.

There are comments in code, but some guidance on reviewing if anyone needs it.
- This uses a no queue pacer when using TWCC (`pacer/no_queue.go`)
- All send packets are queued up by this module and processed in a goroutine, i. e. all downstream packets go through this. This could become a performance bottleneck concern. Have to keep a watch for it. But, with TWCC, would be good to have packets going out in sequence.
- In the forward path, `RecordPacketSendAndGetSequenceNumber` in `sendsidebwe/packet_tracker.go` records the outgoing packet with the TWCC sequence number.
- In the reverse path, `HandleRTCP` in `sendsidebwe/congestion_detector.go` queues up TWCC feedback reports and processes it in a goroutine (`processFeedbackReport` in that file ☝️ ).
- `sendsidebwe/packet_group.go` accumulates packets acknowledged in the TWCC feedback report into groups.
- `sendsidebwe/congestion_detector.go` has congestion detection state machine and bandwidth estimation.
- There are callbacks into `StreamAllocator` which uses the congestion signal to adjust allocation.

TODO:
- Need to work on probing. It is probing now, but it is not long enough.
- Pure loss based congestion is not handled. If there is a middle box that uses dropping as a strategy to deal with congestion rather than building up queue, it is not handled. Need to add a way to deal with it.
- Need to do some code movement (refactor/re-stucturing) to make it cleaner. It is a bit of a tangle now, but lower in priority.

### CODE REVIEW

File: pkg/sfu/rtpstats/rtpstats_sender.go [hunk 1]
```diff
@@ -515,12 +514,10 @@ func (r *RTPStatsSender) UpdateFromReceiverReport(rr rtcp.ReceptionReport) (rtt
 		}
 	}
 
-	// This is 24-bit max in the protocol. So, technically doesn't need extended type. But, done for consistency.
-	packetsLostFromRR := r.packetsLostFromRR&0xFFFF_FFFF_0000_0000 + uint64(rr.TotalLost)
-	if (rr.TotalLost-r.lastRR.TotalLost) < (1<<31) && rr.TotalLost < r.lastRR.TotalLost {
-		packetsLostFromRR += (1 << 32)
+	r.packetsLostFromRR = r.packetsLostFromRR&0xFFFF_FFFF_FF00_0000 + uint64(rr.TotalLost)
+	if ((rr.TotalLost-r.lastRR.TotalLost)&((1<<24)-1)) < (1<<23) && rr.TotalLost < r.lastRR.TotalLost {
+		r.packetsLostFromRR += (1 << 24)
```
Comment (boks1971) [inline pkg/sfu/rtpstats/rtpstats_sender.go:519]: drive-by clean up. There is a 24-bit field in TWCC report also. Was handling that and remembered this. So, doing the same handling here.

**Review by paulwe — APPROVED** [review]
only really have superficial feedback but lgtm 👍

File: pkg/sfu/sendsidebwe/congestion_detector.go [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+
+	estimatedAvailableChannelCapacity int64
+	congestionState                   CongestionState
+	congestionStateSwitchedAt         time.Time
+	congestedCTRTrend                 *ccutils.TrendDetector[float64]
+
+	onCongestionStateChange func(congestionState CongestionState, estimatedAvailableChannelCapacity int64)
+}
+
+func NewCongestionDetector(params congestionDetectorParams) *congestionDetector {
```
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/sfu/sendsidebwe/congestion_detector.go:121]: Yeah true. Will change those. I am always confused by those and I usually end up changing it a few times developing a module :-) . I was thinking if all methods should be private as well, but going to leave those as public, but just change the contructors.

File: pkg/sfu/sendsidebwe/congestion_detector.go [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+func (c *congestionDetector) GetEstimatedAvailableChannelCapacity() int64 {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	return c.estimatedAvailableChannelCapacity
+}
+
+func (c *congestionDetector) HandleRTCP(report *rtcp.TransportLayerCC) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
```
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/sfu/sendsidebwe/congestion_detector.go:202]: have been bitten by "posting to a closed channel" panic enough times :-) . So, tend to lock that as well. But, that is not doing what it is intended here. Will add a check for stop fuse also.

Will `select` not attempt to post if channel is closed? Not able to find good behaviour documentation there.

File: pkg/sfu/sendsidebwe/packet_tracker.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+// -------------------------------------------------------------------------------
+
+type packetTrackerParams struct {
+	Logger logger.Logger
+}
+
+type packetTracker struct {
+	params packetTrackerParams
+
+	lock sync.Mutex
```
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/sfu/sendsidebwe/packet_tracker.go:27]: Unfortunately no. I initially relied on the wrapping lock (from `congestion_detector`), but in the forward path, when sequence numbers are created, `congestion_detector` is not in the path.

Also, in the reverse path, did not want to rely on the wrapping lock as it would mean holding the lock while the report is processed (it could be fast, but it would still affect the forward path).

File: pkg/sfu/sendsidebwe/congestion_detector.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/sfu/sendsidebwe/congestion_detector.go:202]: Will merge this for now. Will change it is posting to channel can be done outside the lock. I think there are other places that can change as well if the posting can happen outside.

File: pkg/sfu/sendsidebwe/congestion_detector.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulwe) [reply pkg/sfu/sendsidebwe/congestion_detector.go]: i think using "public" members on private types can make things more readable it was just he constructors that seemed confusing

File: pkg/sfu/sendsidebwe/congestion_detector.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulwe) [reply pkg/sfu/sendsidebwe/congestion_detector.go:202]: it might make more sense to use `c.stop.Break()` to stop the worker then you don't have to worry about writing to a closed channel. you would never want to close `wake` anyway because as long as `c.stop` isn't broken the worker will infinitely loop selecting the closed wake channel

[the remark this answers is not in the document]
Comment (boks1971) [reply pkg/sfu/sendsidebwe/congestion_detector.go:202]: Sounds good. Will make a small PR to just address that.

File: pkg/sfu/sendsidebwe/packet_tracker.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulwe) [reply pkg/sfu/sendsidebwe/packet_tracker.go:27]: oh, i missed that it's part of the public interface of congestion detector...
